utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2202 @@
+// iterate over the blobs sequentially and check if the blob sequence number
+// is the latest. If it is the latest, preserve it, otherwise delete it
+// if it is TTL based, and the TTL has expired, then
+// we can blow the entity if the key is still the latest or the Key is not
+// found
+// WHAT HAPPENS IF THE KEY HAS BEEN OVERRIDEN. Then we can drop the blob
+// without doing anything if the earliest snapshot is not
+// referring to that sequence number, i.e. it is later than the sequence number
+// of the new key
+//
+// if it is not TTL based, then we can blow the key if the key has been
+// DELETED in the LSM
+////////////////////////////////////////////////////////////////////////////////
+Status BlobDBImpl::GCFileAndUpdateLSM(const std::shared_ptr<BlobFile>& bfptr,
+                                      GCStats* gcstats) {
+  std::chrono::system_clock::time_point now = std::chrono::system_clock::now();
+  std::time_t tt = std::chrono::system_clock::to_time_t(now);
+
+  std::shared_ptr<Reader> reader =
+      bfptr->OpenSequentialReader(myenv_, db_options_, env_options_);
+  if (!reader) {
+    Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+        "File sequential reader could not be opened",
+        bfptr->PathName().c_str());
+    return Status::IOError("failed to create sequential reader");
+  }
+
+  BlobLogHeader header;
+  Status s = reader->ReadHeader(&header);
+  if (!s.ok()) {
+    Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+        "Failure to read header for blob-file %s", bfptr->PathName().c_str());
+    return s;
+  }
+
+  bool first_gc = bfptr->gc_once_after_open_;
+
+  ColumnFamilyHandle* cfh = bfptr->GetColumnFamily(db_);
+  auto cfhi = reinterpret_cast<ColumnFamilyHandleImpl*>(cfh);
+  auto cfd = cfhi->cfd();
+  bool has_ttl = header.HasTTL();
+
+  // this reads the key but skips the blob
+  Reader::ReadLevel shallow = Reader::kReadHdrKeyFooter;
+
+  assert(opt_db_);
+
+  bool no_relocation_ttl = (has_ttl && tt > bfptr->GetTTLRange().second);
+
+  bool no_relocation_lsmdel = false;
+  {
+    ReadLock lockbfile_r(&bfptr->mutex_);
+    no_relocation_lsmdel = (bfptr->GetFileSize() ==
+                            (BlobLogHeader::kHeaderSize + bfptr->deleted_size_ +
+                             BlobLogFooter::kFooterSize));
+  }
+
+  bool no_relocation = no_relocation_ttl || no_relocation_lsmdel;
+  if (!no_relocation) {
+    // read the blob because you have to write it back to new file
+    shallow = Reader::kReadHdrKeyBlobFooter;
+  }
+
+  BlobLogRecord record;
+  std::shared_ptr<BlobFile> newfile;
+  std::shared_ptr<Writer> new_writer;
+
+  while (reader->ReadRecord(&record, shallow).ok()) {
+    gcstats->blob_count++;
+
+    bool del_this = false;
+    // this particular TTL has expired
+    if (no_relocation_ttl || (has_ttl && tt > record.GetTTL())) {
+      del_this = true;
+    } else {
+      SequenceNumber seq = kMaxSequenceNumber;
+      bool found_record_for_key = false;
```

Comment:
I feel like the transaction should start here, not below. Otherwise there still can be data race that we miss.